### DISCUSSION TITLE
Sticky Menu Toggling Smoothly on Scroll - JavaScript Code For systeme.io Pages

### DISCUSSION THREAD

**Discussion by sofiane-abou-abderrahim** [community/community#64316] [Programming Help]
### Body

I need a JavaScript code to implement in my systeme.io pages that is able to make a sticky menu toggling smoothly on scroll.

The sticky menu should have an initial state and initial styles at the initial load of the page.

Then, when scrolling down the stcky menu toggles smoothly and the logo's color, the background color, the menu's links and the button's background color change.

And Finally, when we scroll up and reach the top of the page, the sticky menu reset smoothly to its initial styles and initial state.

Here two example of what I would like to achieve:

[https://www.make.com/en](https://www.make.com/en)

[https://impact2020.edf.org/](https://impact2020.edf.org/)

Here is my code:

```
<script>
  const logoImage = document.getElementById('image-c418c30e');
  const whiteLogoUrl = 'https://d1yei2z3i6k35z.cloudfront.net/4555433/64dc0781ef611_sticky-menu-white-logo.png';
  const coloredLogoUrl = 'https://d1yei2z3i6k35z.cloudfront.net/4555433/64dc0776841e4_sticky-menu-colored-logo.png';

  const coloredLogo = new Image();
  coloredLogo.src = coloredLogoUrl;

  const whiteLogo = new Image();
  whiteLogo.src = whiteLogoUrl;

  logoImage.src = whiteLogoUrl;

  document.addEventListener('DOMContentLoaded', function () {
    const stickyMenu = document.getElementById('section-4f6994fa');
    const menu = document.getElementById('menu-972c4b46');
    const button = document.getElementById('button-e246b912');
    const navigationLinks = menu.querySelectorAll('a');

    const initialStyles = {
      backgroundImage: stickyMenu.style.backgroundImage,
      backgroundColor: stickyMenu.style.backgroundColor,
      menuColor: menu.style.color,
      buttonColor: button.style.backgroundColor
    };

    const transitionDuration = '0.5s';
    
    menu.style.transition = `color ${transitionDuration} ease-in-out`;
    navigationLinks.forEach(link => {
      link.style.transition = `color ${transitionDuration} ease-in-out`;
    });
    button.style.transition = `background-color ${transitionDuration} ease-in-out`;
    logoImage.style.transition = `opacity ${transitionDuration} ease-in-out`;
    stickyMenu.style.transition = `background-color ${transitionDuration} ease-in-out`; // Faster transition

    let isAtTop = true;
    let isButtonBlack = false;
    let logoSwapped = false;
    let isTransitioning = false;

    function swapLogo(sourceUrl) {
      logoImage.style.opacity = '0';
      setTimeout(() => {
        logoImage.src = sourceUrl;
        logoImage.style.opacity = '1';
        logoSwapped = false;
      }, 300);
    }

    function smoothTransition(element, property, fromValue, toValue, duration) {
      const fps = 60;
      const interval = 1000 / fps;
      const steps = duration / interval;
      let stepCount = 0;
      
      const increment = (toValue - fromValue) / steps;

      const timer = setInterval(() => {
        if (stepCount >= steps) {
          clearInterval(timer);
          element.style[property] = toValue;
        } else {
          element.style[property] = `${fromValue + increment * stepCount}px`;
          stepCount++;
        }
      }, interval);
    }

function resetStyles(styles) {
  // Use a faster duration for the background color reset transition
  stickyMenu.style.transition = `background-color 0.3s ease-in-out`; // Add this line
  stickyMenu.style.backgroundColor = 'transparent'; // Set the background color to transparent
  setTimeout(() => {
    stickyMenu.style.backgroundImage = styles.backgroundImage;
    stickyMenu.style.backgroundColor = styles.backgroundColor;
    stickyMenu.style.transition = 'background-color 0.5s ease-in-out'; // Reset the transition
  }, 300); // Same timing as the background color transition

      // Reset other styles here
      menu.style.color = styles.menuColor;
      navigationLinks.forEach(link => {
        link.style.color = styles.menuColor;
      });
      button.style.backgroundColor = styles.buttonColor;
    }

function toggleBackgroundOnScroll(elementId) {
    const section = document.getElementById(elementId);
    if (!section) {
      console.error(`Element with ID "${elementId}" not found.`);
      return;
    }

    let prevScrollPos = window.scrollY;

    function handleScroll() {
      const currentScrollPos = window.scrollY;

      if (currentScrollPos === 0) {
        resetStyles(initialStyles);
        if (!logoSwapped) {
          swapLogo(whiteLogoUrl);
          logoSwapped = true;
        }
        isAtTop = true;
        if (isButtonBlack) {
          button.style.backgroundColor = initialStyles.buttonColor;
          isButtonBlack = false;
        }
        navigationLinks.forEach((link, index) => {
          link.style.color = initialStyles.menuColor;
        });
      } else if (currentScrollPos > 0 && currentScrollPos > prevScrollPos && isAtTop) {
        if (!isTransitioning) {
          isTransitioning = true;
          section.style.backgroundColor = 'white';
          menu.style.color = 'black';
          navigationLinks.forEach(link => {
            link.style.color = 'black';
          });
          button.style.backgroundColor = 'black';
          setTimeout(() => {
            section.style.backgroundImage = `url(${coloredLogoUrl})`;
            isTransitioning = false;
          }, 150); // Same timing as swapping the logo
        }
        if (!logoSwapped) {
          swapLogo(coloredLogoUrl);
          logoSwapped = true;
        }
        isAtTop = false;
      } else if (currentScrollPos < prevScrollPos && !isAtTop) {
        menu.style.color = initialStyles.menuColor;
        if (!isAtTop && currentScrollPos === 0) {
          if (logoSwapped) {
            swapLogo(whiteLogoUrl);
            logoSwapped = false;
          }
          button.style.backgroundColor = initialStyles.buttonColor;
          isButtonBlack = false;
          navigationLinks.forEach((link, index) => {
            link.style.color = initialStyles.menuColor;
          });
        }
      }

      prevScrollPos = currentScrollPos;
    }

    window.addEventListener('scroll', handleScroll);
  }

  toggleBackgroundOnScroll('section-4f6994fa');

  });
</script>
```

As you can see, the code targets several elements of the page by IDs and URLs. For example, the sticky menu has the ID 'section-4f6994fa'. The logos swap between a colored logo and a white logo targeted by URLs.

This code kind of works, but there are few problems with it:

The sticky menu does not toggles smoothly anough. It is too abrupt, especially at the end of the transition.

This code doesn't work with all my systeme.io's pages. For example, I had to remove the line 121 in other pages for it to work:

```
section.style.backgroundImage = `url(${coloredLogoUrl})`;
```

When, I leave this line in the other page, the sticky menu's background color changes to white, it displays a big image of the colored logo behind, besides the logo image. And when I remove this line of the code, the big image of the colored logo disappears and the code works.

However, when I remove this line of code in another systeme.io's page, the sticky menu's background color doesn't transition to white at all.

Indeed, I need a general code, like a template, that would work in every systeme.io's page. I would just have to replace the IDs and URLs.

### Guidelines

- [X] I have read and understood this category's [guidelines](https://github.com/orgs/community/discussions/51384) before making this post.

**Comment by yahiaelbanna** [comment]
You are adding a lot of lines in your javascript code when you can simplify your code by just changing a class name when the user scrolls down!
when you depend on javascript files only then you are taking a risk with your website performance!

All these lines of code can be converted into this code :
`document.addEventListener('scroll',_=>{
            window.scrollY > 90 ? document.querySelector('#header').classList.add('scrolled') : document.querySelector('#header').classList.remove('scrolled');
        })`

and with just a little bit of CSS coding, you can have an amazing result!
you can check this [pen](https://codepen.io/yahiaelbnna/pen/jOXEZje) I have created to show you the new code.

**Reply by sofiane-abou-abderrahim** [reply]
_you can check this [https://codepen.io/yahiaelbnna/pen/jOXEZje](https://github.com/orgs/community/discussions/pen) I have created to show you the new code_.

Your link is redirecting to 404 error.

I have to write one code to integrate it to systeme.io's page builder. So, if there is CSS code, it should be combined to this JavaScript code.
If the method you described in your reply can do so, I would like you to show it to me please.

**Reply by yahiaelbanna** [reply]
This link should direct you to codepen.io .. you can open it now.

now I understood your request. I have created a javascript library that can help you with your request. if you like I will code your request with it and send you the repository link.

you can find this library here: [gmsmpl](https://github.com/yahaielbnna/gmsmpl) until I code your request.

**Reply by yahiaelbanna** [reply]
You now can check this code and hope this is helpful to you
[code](https://github.com/yahaielbnna/github-community/tree/main/sticky-menu-toggling-on-scroll)

**Reply by sofiane-abou-abderrahim** [reply]
I checked your code and it works in my browser.

However, as I said earlier, I have to integrate this code in my syseme.io's pages.
For this, I need to insert a piece of code directly in the systeme.io's page builder.

So, how can I use your code in this way?

Usually, like you can see in my initial code, we have to use IDs to target the elements in systeme.io.
For example, the sticky menu has an ID, the logo image has an ID, the menu has an ID, the button has an ID.
You can see those IDs in my code.
Then, where can I put those IDs in your code?

Therefore, I can't use an HTML file and link css files or JavaScript files. Everything, should be in the same code.

If you understood what I'm talking about, please show me the way.
Besides that, please explain me your code in detail, because it is a quiet big code.

**Reply by yahiaelbanna** [reply]
you can read the comment beside every line that must help you. 
[the code](https://github.com/yahaielbnna/github-community/tree/main/sticky-menu-toggling-on-scroll)

to make it much clearer. The HTML code just has one element `<header id="header"></header>`. also, there aren't any CSS files.
script.js file works to create the new element and add styles to it using the "gmsmpl" library.  
as you can see the functions named with CSS's property and there are some functions like (addAttr=> adding attributes, setId=> set id to the element, $ => query select, parent => select the selected element parent, insert => creating a new elements and any functions calls after that work with the new element)

**Reply by sofiane-abou-abderrahim** [reply]
Okay, thank you for your reply.

So, how can I integrate your code to my systeme.io's page via the systeme.io's page builder?

It has to be one code that targets the elements on which we want to make an animation by IDs.

For example, I think that the header should be the section with the ID 'section-4f6994fa'.

Please show me the way I can use your code and adapt it to my systeme.io's page.

**Reply by yahiaelbanna** [reply]
actually, I don't know what is  " systeme.io ". but as you said It should be a Javascript file. so I used "gmsmpl" to create all the elements and style them without using any CSS files.

if you want to select any element you can use this function "`$(' -- Query select -- ')`", then you start to style it or insert elements into it and any function from the available in  "gmsmpl " until now. 

I just created an example that shows you how to use this library. If you see it can help you then you should start to use it for your needs.

**Comment by proksh** [comment]
To achieve this, let's:
- Simplify and improve the transition to make it smoother.
- Remove any hard-coded element-specific styles to make it more general.

Here's a refactored version of your code:
```
<script>
  document.addEventListener('DOMContentLoaded', function () {
    const logoImage = document.getElementById('image-c418c30e');
    const stickyMenu = document.getElementById('section-4f6994fa');
    const menu = document.getElementById('menu-972c4b46');
    const button = document.getElementById('button-e246b912');
    const navigationLinks = menu.querySelectorAll('a');

    // Logo URLs
    const whiteLogoUrl = 'https://d1yei2z3i6k35z.cloudfront.net/4555433/64dc0781ef611_sticky-menu-white-logo.png';
    const coloredLogoUrl = 'https://d1yei2z3i6k35z.cloudfront.net/4555433/64dc0776841e4_sticky-menu-colored-logo.png';

    // Initial styles
    const initialStyles = {
      backgroundImage: getComputedStyle(stickyMenu).backgroundImage,
      backgroundColor: getComputedStyle(stickyMenu).backgroundColor,
      menuColor: getComputedStyle(menu).color,
      buttonColor: getComputedStyle(button).backgroundColor
    };

    const transitionDuration = '0.5s';
    const applyTransitions = (element, properties) => {
      properties.forEach(prop => {
        element.style.transition += `${prop} ${transitionDuration} ease-in-out, `;
      });
    };

    applyTransitions(menu, ['color']);
    applyTransitions(button, ['background-color']);
    applyTransitions(stickyMenu, ['background-color']);
    navigationLinks.forEach(link => {
      applyTransitions(link, ['color']);
    });

    let isAtTop = true;

    function swapLogo(sourceUrl) {
      logoImage.style.opacity = '0';
      setTimeout(() => {
        logoImage.src = sourceUrl;
        logoImage.style.opacity = '1';
      }, 500 * 0.6); // 60% of transition duration
    }

    function resetStyles(styles) {
      stickyMenu.style.backgroundColor = styles.backgroundColor;
      menu.style.color = styles.menuColor;
      navigationLinks.forEach(link => {
        link.style.color = styles.menuColor;
      });
      button.style.backgroundColor = styles.buttonColor;
    }

    function handleScroll() {
      const currentScrollPos = window.scrollY;

      if (currentScrollPos === 0 && !isAtTop) {
        resetStyles(initialStyles);
        swapLogo(whiteLogoUrl);
        isAtTop = true;
      } else if (currentScrollPos > 0 && isAtTop) {
        stickyMenu.style.backgroundColor = 'white';
        menu.style.color = 'black';
        navigationLinks.forEach(link => {
          link.style.color = 'black';
        });
        button.style.backgroundColor = 'black';
        swapLogo(coloredLogoUrl);
        isAtTop = false;
      }
    }

    window.addEventListener('scroll', handleScroll);
  });
</script>
```

What was changed/optimized:

- Moved the logo URLs to the top of the DOMContentLoaded block for clarity.
- Simplified the transition setting by creating an applyTransitions function.
- Used getComputedStyle to retrieve initial styles, which is more reliable.
- Removed unnecessary checks and redundant code to simplify the scroll handler.
- Removed the hard-coded element-specific style changes, so it's more generalized.

To use this code on any systeme.io page:

- Update the element IDs ('image-c418c30e', 'section-4f6994fa', etc.) to match the elements on that specific page.
- Ensure that the initial styles (background image, colors, etc.) are set using CSS so that they can be fetched using getComputedStyle.

The above refactored code should provide a smoother experience and be more flexible across your different pages.

**Reply by sofiane-abou-abderrahim** [reply]
I tried your code and the sticky menu's behaviour changed. It doesn't have the same result. It should have the same result as my code.
